City-of-Helsinki/atv
# ATV-73 | API authentication with a Tunnistamo OIDC token

## Description :sparkles:

This PR adds necessary configuration and packages for API authentication using [tunnistamo](https://github.com/City-of-Helsinki/tunnistamo/). User is automatically created based on the token information.

## Issues :bug:
### Closes :no_good_woman:
**[ATV-73](https://helsinkisolutionoffice.atlassian.net/browse/ATV-73):** Accept Tunnistamo OIDC token as authentication

## Testing :alembic:

### Manual testing :construction_worker_man:

For accessing a the API using a JWT/OIDC token, you need to request one from Tunnistamo.

Setup local Tunnistamo (or use some hosted one):

- Add a client
  - Client id: https://api.hel.fi/auth/atv
  - Name: ATV
  - Redirect URI: https://oidcdebugger.com/debug
  - Login methods e.g. github
  - Response types: id_token token

- Add an API
  - Domain: https://api.hel.fi/auth
  - Name: atv
  - OIDC client: atv

- Add an API scope
  - API: https://api.hel.fi/auth/atv
  - name: atv
  - specifier: (leave empty)
  - description atv

Once Tunnistamo has been setup, you should be able to use the following environment variables for configuration configuration:

```
TOKEN_AUTH_ACCEPTED_SCOPE_PREFIX=atv
TOKEN_AUTH_ACCEPTED_AUDIENCE=https://api.hel.fi/auth/atv
TOKEN_AUTH_AUTHSERVER_URL=http://tunnistamo-backend:8000/openid
TOKEN_AUTH_REQUIRE_SCOPE=True
```

1. Request a token with: `http://tunnistamo-backend:8000/openid/authorize?client_id=https://api.hel.fi/auth/atv&redirect_uri=https://oidcdebugger.com/debug&scope=openid https://api.hel.fi/auth/atv&response_type=id_token token&response_mode=form_post&nonce=hello4`
2. Fetch an API token using the token from previous step as a `Bearer` token.
http://tunnistamo-backend:8000/api-tokens
3. Use the API token in the API


## Additional notes :spiral_notepad:

Requires [tunnistamo](https://github.com/City-of-Helsinki/tunnistamo/)
